runtime/tests/integration_test.rs

```diff
@@ -0,0 +1,303 @@
+	ExtBuilder::default()
+		.with_balances(vec![(AccountId::from(ALICE), 3_000 * GLMR)])
+		.build()
+		.execute_with(|| {
+			// Charlie has no balance => fails to stake
+			assert_noop!(
+				ParachainStaking::join_candidates(
+					origin_of(AccountId::from(CHARLIE)),
+					1_000 * GLMR,
+				),
+				DispatchError::Module {
+					index: 3,
+					error: 3,
+					message: Some("InsufficientBalance")
+				}
+			);
+			// Alice stakes to become a collator candidate
+			assert_ok!(ParachainStaking::join_candidates(
+				origin_of(AccountId::from(ALICE)),
+				1_000 * GLMR,
+			));
+			// Alice transfer from free balance 1000 GLMR to Bob
+			assert_ok!(Balances::transfer(
+				origin_of(AccountId::from(ALICE)),
+				AccountId::from(BOB),
+				2_000 * GLMR,
+			));
+			assert_eq!(Balances::free_balance(AccountId::from(BOB)), 2_000 * GLMR,);
+			use sp_core::U256;
+			let gas_limit = 100000u64;
+			let gas_price: U256 = 1000.into();
+			// Bob transfers 1000 GLMR to Charlie via EVM
+			assert_ok!(Call::EVM(pallet_evm::Call::<Runtime>::call(
+				AccountId::from(BOB),
+				AccountId::from(CHARLIE),
+				Vec::new(),
+				(1_000 * GLMR).into(),
+				gas_limit,
+				gas_price,
+				None
+			))
+			.dispatch(origin_of(AccountId::from(BOB))));
+			assert_eq!(
+				Balances::free_balance(AccountId::from(CHARLIE)),
+				1_000 * GLMR,
+			);
+			// Charlie can stake now
+			assert_ok!(ParachainStaking::join_candidates(
+				origin_of(AccountId::from(CHARLIE)),
+				1_000 * GLMR,
+			),);
+		});
+}
+
+#[test]
+fn reward_block_authors() {
+	ExtBuilder::default()
+		.with_balances(vec![
+			(AccountId::from(ALICE), 2_000 * GLMR),
+			(AccountId::from(BOB), 1_000 * GLMR),
+		])
+		.with_collators(vec![(AccountId::from(ALICE), 1_000 * GLMR)])
+		.with_nominators(vec![(
+			AccountId::from(BOB),
+			AccountId::from(ALICE),
+			500 * GLMR,
+		)])
+		.build()
+		.execute_with(|| {
+			// set parachain inherent data
+			use cumulus_primitives_core::PersistedValidationData;
+			use cumulus_test_relay_sproof_builder::RelayStateSproofBuilder;
+			let (relay_parent_storage_root, relay_chain_state) =
+				RelayStateSproofBuilder::default().into_state_root_and_proof();
+			let vfp = PersistedValidationData {
+				relay_parent_number: 1u32,
+				relay_parent_storage_root,
+				..Default::default()
+			};
+			let parachain_inherent_data = ParachainInherentData {
+				validation_data: vfp,
+				relay_chain_state: relay_chain_state,
+				downward_messages: Default::default(),
+				horizontal_messages: Default::default(),
+			};
+			assert_ok!(Call::ParachainSystem(
+				cumulus_pallet_parachain_system::Call::<Runtime>::set_validation_data(
+					parachain_inherent_data
+				)
+			)
+			.dispatch(inherent_origin()));
+			fn set_alice_as_author() {
+				assert_ok!(
+					Call::AuthorInherent(author_inherent::Call::<Runtime>::set_author(
+						AccountId::from(ALICE)
+					))
+					.dispatch(inherent_origin())
+				);
+			}
+			for x in 2..1201 {
+				set_alice_as_author();
```

Comment:
why do you need to set alice as author if she is the only collator? Maybe set up a test with more than one collator and somehow count how many time each collator was selected as author and test that the balance reward is correct in a dynamic way

Reply:
An author must be set every block or `on_finalize` in `author-inherent` panics. 

Having multiple collators to test correct reward distribution is already done in `parachain-staking` unit tests. I think it's reasonable to add the same test here as well.

Reply:
@joelamouche This is a good chance to learn about Substrate inherents. When a collator authors a block, they have the opportunity to include inherent extrinsics. In moonbeam there are three of them (timestamp, validation_data, and author) and all three are required. It is behind-the-scenes from a user perspective, but every time a collator authors a moonbeam block, they insert these three inherents. Since Amar is mocking here, and there is no collator process to do any actual block authoring, we have to call these inherents manually. So setting the author to Alice here _is_ mocking Alice authoring the block.

Reply:
> Maybe set up a test with more than one collator and somehow count how many time each collator was selected as author and test that the balance reward is correct in a dynamic way

I'd rather do this at the pallet level  (and have done it there) where I can mock the `set_author` command in a more deterministic way. I added Charlie, Dave, and Eve to the set of collators and none of them were included in the set of valid block authors so it kept returning the following error:

```
---- reward_block_authors stdout ----
thread 'reward_block_authors' panicked at 'Expected Ok(_). Got Err(
    DispatchErrorWithPostInfo {
        post_info: PostDispatchInfo {
            actual_weight: None,
            pays_fee: Pays::Yes,
        },
        error: DispatchError::Module {
            index: 17,
            error: 1,
            message: Some(
                "CannotBeAuthor",
            ),
        },
    },
)', runtime/tests/integration_test.rs:337:17
```

Testing multiple block authors in a runtime integration test is worthwhile, but I'd prefer for it to be pursued in a follow up because I have other things that I need to focus on (staking precompile, benchmarking).